### PR TITLE
WIP: Add initial server/backend implementation

### PR DESCRIPTION
**Pull request by RauliL**
Add server/backend classes and launcher script for running HTTP/WebSocket server that serves games in VS mode.

**Not mergeable**: Currently I receive "State cache corrupted" errors with this.

### CODE REVIEW

File: lib/game.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      this.players.splice(index, 1);
+      // TODO: Signal remaining players about the disconnected player.
+      if (!this.players.length) {
+        this.server.removeGame(this);
+      }
+    }
+  }
+
+  addEvent(data, source) {
+    this.engine.addEvent(data);
```
Comment (frostburn) [inline lib/game.js]: `this.engine.addEvent(data.event);` instead.

File: lib/game.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    });
+  }
+
+  addPlayer(player) {
+    if (this.players.length > 1) {
+      return;
+    }
+    player.game = this;
+    this.players.push(player);
+    if (this.players.length === 2) {
```
Comment (frostburn) [inline lib/game.js]: The `2` here should be a module level constant and should be added to options as `numPlayers` to be explicit about the defaults.

File: lib/game.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      process.stdout.write('Game is starting...\n');
+      this.players.forEach((player, index) => {
+        player.socket.emit('connected', {
+          player: index,
+          game: this.engine.serialize(),
+          frameRate: this.server.frameRate,
+          cache: this.engine.exportCache(),
```
Comment (frostburn) [inline lib/game.js]: Cache exporting is already handled by `this.engine.serialize()`.

File: lib/game.js (hunk 2 above)
Comment (RauliL) [inline lib/game.js]: Actually I was thinking that different game modes would be objects and maximum number of players would be read from them.

**Comment by frostburn** [comment]
I'll take it from here.

**Comment by frostburn** [comment]
Obsoleted by https://github.com/frostburn/panel-league/pull/68

**Pull request closed without merging**